### PR TITLE
[Constraint system] Eliminate global flag UnderlyingTypeForOpaqueReturnType

### PR DESCRIPTION
**Pull request by DougGregor**
Make this information contextual (per type) rather than global (per
constraint system) so we don’t misapply it.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
Only the last two commits are unmerged. Huh.

File: lib/Sema/CSSolver.cpp [hunk 1]
```diff
@@ -1115,80 +1116,93 @@ bool ConstraintSystem::solve(Expr *&expr,
...
-      .highlight(expr->getSourceRange());
-    return true;
-  }
+      if (Options.contains(
+            ConstraintSystemFlags::AllowUnresolvedTypeVariables)) {
+        auto ambiguousSolutions = std::move(solution).takeAmbiguousSolutions();
+        solutions.assign(std::make_move_iterator(ambiguousSolutions.begin()),
+                         std::make_move_iterator(ambiguousSolutions.end()));
+        dumpSolutions();
+        solution.markAsDiagnosed();
```
Comment (xedin) [inline lib/Sema/CSSolver.cpp]: Looks like the same 'use-after-move' problem @CodaFi mentioned is present here as well...

**Review by xedin — APPROVED** [review]
Looks good to me, it does make more sense to have opaque result bit as a contextual information instead of a flag.

File: lib/Sema/CSSolver.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/CSSolver.cpp]: Hmm. It's the same one. I'll roll this into an upcoming PR.

**Pull request merged**